File: 0707_Design Linked List/DesignLinkedList.java
class MyLinkedList {
    class Node {
      public int val;
      public Node next;
      public Node(int val) { this.val = val; this.next = null; }
      public Node(int val, Node next) { this.val = val; this.next = next; }
    }
    
    private Node _head;
    private Node _tail;
    private int _size;
    
    public MyLinkedList() {
      this._head = this._tail = null;
      this._size = 0;
    }
    
    private Node getNode(int index) {
      Node node = new Node(0, this._head);
      while (index-- >= 0) {
        node = node.next;
      }
      return node;
    }
   
    public int get(int index) {
      if (index < 0 || index >= _size) return -1;
      return getNode(index).val;
    }
   
    public void addAtHead(int val) {
      this._head = new Node(val, this._head);
      if (this._size++ == 0)
        this._tail = this._head;    
    }
   
    public void addAtTail(int val) {
      Node node = new Node(val);
      if (this._size++ == 0)
        this._head = this._tail = node;
      else
        this._tail = this._tail.next = node;
    }
   
    public void addAtIndex(int index, int val) {
      if (index < 0 || index > this._size) return;
      if (index == 0)  { this.addAtHead(val); return; }
      if (index == _size) { this.addAtTail(val); return; }
      Node prev = this.getNode(index - 1);
      prev.next = new Node(val, prev.next);
      ++this._size;
    }
   
    public void deleteAtIndex(int index) {
      if (index < 0 || index >= this._size) return;
      Node prev = this.getNode(index - 1);
      prev.next = prev.next.next;
      if (index == 0) this._head = prev.next;
      if (index == this._size - 1) this._tail = prev;
      --this._size;
    }
  }

File: 0707_Design Linked List/DesignLinkedList.py
class Node:
  def __init__(self, val, _next=None):
    self.val = val
    self.next = _next
    
class MyLinkedList:  
  def __init__(self):
    self.head = self.tail = None
    self.size = 0
  
  def getNode(self, index):
    node = Node(0, self.head)
    for i in range(index + 1):
      node = node.next
    return node
    
  def get(self, index):
    if index < 0 or index >= self.size: return -1
    return self.getNode(index).val
 
 
  def addAtHead(self, val):
    node = Node(val, self.head)
    self.head = node
    if self.size == 0:
      self.tail = node
    self.size += 1
 
 
  def addAtTail(self, val):
    node = Node(val)
    if self.size == 0:
      self.head = self.tail = node
    else:
      self.tail.next = node
      self.tail = node
    self.size += 1
 
  def addAtIndex(self, index, val):
    if index < 0 or index > self.size: return
    if index == 0: return self.addAtHead(val)
    if index == self.size: return self.addAtTail(val)
    prev = self.getNode(index - 1)
    node = Node(val, prev.next)
    prev.next = node
    self.size += 1
 
 
  def deleteAtIndex(self, index):
    if index < 0 or index >= self.size: return
    prev = self.getNode(index - 1)
    prev.next = prev.next.next
    if index == 0: self.head = prev.next
    if index == self.size - 1: self.tail = prev
    self.size -= 1